### PR TITLE
Integrate teams and contexts

### LINKED ISSUE #28: Resolve context-team interaction issues
The current draft of the teams proposal puts the PE team inside the communication context and allows for the reassignment of teams to contexts. This issue is meant to detail the proposed model and potential alternatives and capture feedback or issues with any of these models as well as other interactions between teams and contexts.

## Potential models for interaction

1. <a href="#TiC">Teams in Contexts</a>
2. <a href="#CiT">Contexts in Teams</a>
3. <a href="#PlanB">Teams use `SHMEM_CTX_DEFAULT`</a>
4. <a href="#CfT">Contexts from Teams</a>

*****

### <a name="TiC">Teams in Contexts</a>

- Each context is associated with a team.
- Each team may be assigned to zero or more contexts.
- By default (i.e., at context creation), each context is associated with `SHMEM_TEAM_WORLD`.
- A context may have its associated team updated (e.g., via `shmem_ctx_set_team`).

#### Potential advantages

- Context-based point-to-point operations can use team-relative PE numbering.
- Updating the team associated to a context could be as simple as a pointer assignment.
- Contexts represent semi-virtualized hardware resources. Separating context and team creation and allowing the reassignment of teams to context, could minimize underutilized resources.

#### Disadvantages or concerns

- From https://github.com/gmegan/specification/issues/28#issuecomment-401832971:
  - Networks may need coordinated source- and target-side resources to enable full messaging throughput (e.g., in multi-HCA/HFI setups).
  - OpenSHMEM implementations that share lower implementation layers with MPI or ones that seek to provide MPI interoperability may want to share resources and implementation paths, potentially backing a SHMEM team with an MPI communicator.
  - Networks with hardware collective acceleration may need these resources to be allocated collectively.
- Concurrent collective calls from multiple threads may not be possible (e.g., any internal `pSync` or `pWork` state is per-PE, not per-thread).

*****

### <a name="CiT">Contexts in Teams</a>

- Each team is associated with a context.
- Each context may be assigned to zero or more contexts.
- At team creation, the associated context is provided and fixed.

#### Potential advantages

- Intended to address the concerns of "Teams in Contexts"
- Could support concurrent collective calls (e.g., each thread has a team, so internal state is thread/team-specific).

#### Disadvantages or concerns

- Generally a more-restrictive model.
- May lead to creation of more teams (i.e., consumption of more resources) than with Teams-in-Contexts. For example, a software communication pipeline that wants to use _N_ contexts on one logical team will need to create _N_ team instances of the same logical team.
- Team-relative point-to-point operations would require a new `shmem_team_*`-prefixed API or manual PE index translation.

*****

### <a name="PlanB">Teams use `SHMEM_CTX_DEFAULT` ("Plan B")</a>

- Teams and contexts are created and managed separately.
- All teams are implicitly associated with the default context (e.g., `SHMEM_CTX_DEFAULT`).
- Team-based collective operations added.
- Context-based point-to-point operations exist as-is (i.e., use "world" PE numbering); no team-based point-to-point operations are added.

#### Disadvantages or concerns

- No opportunity for communication concurrency in collectives.
- Team-relative point-to-point operations require manual PE index translation.

*****

### <a name="CfT">Context _from_ Teams</a> (updated 7/23)

- Given a team, a team-specific context can be created (e.g., `shmem_team_create_ctx`).
- Concurrent collectives on the same _logical_ team require distinct team objects.
- Operations on contexts are performed relative to the associated team. The default context and local semantic contexts (`shmem_ctx_create`) are associated with `SHMEM_TEAM_WORLD`.

#### Use Model

- Program creates a new team: `shmem_team_split_*` (collective operation)
  - Internal `pSync` and `pWrk` state lives with the team.
  - Depending on the network, hardware offload engine state may live with the team.
- Program creates team-based context(s) from that team: `shmem_team_create_ctx` (collective operation)
  - Depending on the network, hardware offload engine state may live with the context.
- Program uses team-based context(s), local-semantic contexts (`shmem_ctx_create`), or the default context for RMA, AMO, collective, and synchronization operations
  - P2P operations are performed with team-based PE numbering. (`SHMEM_TEAM_WORLD` is the associated team for all local-semantic contexts and the default context.)
  - Collective operations may be performed on any context, however, the team associated with the context is the point of serialization.
- Program destroys team-based context(s): `shmem_team_destroy_ctx` (collective operation)
- Program destroys the team: `shmem_team_destroy` (collective operation)

#### Potential advantages

- Allows for collective wire-up of any required resources.
- Allows for communication concurrency with concurrent collectives (using distinct team objects)

#### Disadvantages or concerns

- ???

#### Open questions + design points

- What is the "final" handle for team-relative operations (including RMA, AMO, and collectives)?
  - ~~If it is the team, then...~~
  - If it is the context, then...
    - ~~Using distinct context handles -- even with each created from `shmem_team_create_ctx` -- may not be safe for concurrent collectives. They would need to be distinct contexts created from distinct teams.~~ Not a _concern_; this is consistent with the requirement for global serialization of collectives using the same active set in 1.4.

### PR DESCRIPTION
**Pull request by nspark**
This PR integrates teams and contexts as discussed in Issue #28.

- Initial team-related context intro text
- Adds:
  - `shmem_team_config_t` configuration parameters
  - `shmem_team_create_ctx`
  - `shmem_team_sync`
- Updates `shmem_team_split_*` for config params
  - Includes changes for PR #23 (3D &#8594; 2D)

Closes #28

### CODE REVIEW

**Comment by nspark** [comment]
[Draft PDF as of `62feca0`](https://github.com/gmegan/specification/files/2298599/openshmem-draft-62feca0.pdf)

Poke @gmegan @shamisp

**Comment by nspark** [comment]
[Draft PDF as of `a91ed96`](https://github.com/gmegan/specification/files/2298983/openshmem-draft-a91ed96.pdf)

**Comment by shamisp** [comment]
[teams overview.pdf](https://github.com/gmegan/specification/files/2308761/teams.overview.pdf)

File: content/shmem_barrier.tex [hunk 1]
```diff
@@ -79,6 +79,14 @@
     Calls to \FUNC{shmem\_ctx\_quiet} can be performed prior
     to calling the barrier routine to ensure completion of operations issued on
     additional contexts.
+
+    \newtext{
+    No team-based barrier is provided by \openshmem, as a team may have any
+    number of communication contexts associated with the team.
```
Comment (gmegan) [inline content/shmem_barrier.tex:85]: with that team

**Comment by shamisp** [comment]
@nspark do we want to merge this in and then submit PRs on top of this one ?

**Comment by gmegan** [comment]
That sounds like a good idea to me. Looking through here, it all looks like it is reasonable resolution of the discussion. I would suggest some changes to the way config options are set, but that seems better left to another issue.

**Pull request merged**